### PR TITLE
bpo-22559: Implementation of the PEP 546: Backport MemoryBIO

### PR DESCRIPTION
**Pull request by vstinner**
Implementation of the PEP 546: backport ssl.MemoryBIO and ssl.SSLObject from master to the 2.7 branch.

Co-Authored-By: Alex Gaynor <alex.gaynor@gmail.com>

http://bugs.python.org/issue22559

https://bugs.python.org/issue22559

### CODE REVIEW

**Comment by vstinner** [comment]
cc @lukasa

File: Doc/library/ssl.rst [hunk 1]
```diff
@@ -1688,6 +1774,143 @@ to be aware of:
...
+   higher level API. It polls for events using the :mod:`selectors` module and
+   handles :exc:`SSLWantWriteError`, :exc:`SSLWantReadError` and
+   :exc:`BlockingIOError` exceptions. It runs the SSL handshake asynchronously
+   as well.
+
+
+Memory BIO Support
+------------------
+
+.. versionadded:: 2.7.9
```
Comment (vstinner) [inline Doc/library/ssl.rst]: Oops, it will be 2.7.14.

File: Doc/library/ssl.rst [hunk 2]
```diff
@@ -1688,6 +1774,143 @@ to be aware of:
...
+   - :meth:`~SSLSocket.read`
+   - :meth:`~SSLSocket.write`
+   - :meth:`~SSLSocket.getpeercert`
+   - :meth:`~SSLSocket.selected_npn_protocol`
+   - :meth:`~SSLSocket.cipher`
+   - :meth:`~SSLSocket.compression`
+   - :meth:`~SSLSocket.pending`
+   - :meth:`~SSLSocket.do_handshake`
+   - :meth:`~SSLSocket.unwrap`
+   - :meth:`~SSLSocket.get_channel_binding`
```
Comment (vstinner) [inline Doc/library/ssl.rst:1820]: selected_alpn_protocol() should be documented too

File: Lib/ssl.py [hunk 3]
```diff
@@ -521,6 +527,136 @@ def _https_verify_certificates(enable=True):
         _create_default_https_context = _create_unverified_context
 
 
+
```
Comment (vstinner) [inline Lib/ssl.py]: PEP8 nitpick: too many empty lines ;-)

File: Modules/_ssl.c [hunk 4]
```diff
@@ -553,10 +568,21 @@ newPySSLSocket(PySSLContext *sslctx, PySocketSockObject *sock,
...
     self->ssl = NULL;
     self->Socket = NULL;
-    self->ssl_sock = NULL;
     self->ctx = sslctx;
     self->shutdown_seen_zero = 0;
     self->handshake_done = 0;
+    self->owner = NULL;
+    if (server_hostname != NULL) {
+        hostname = PyUnicode_Decode(server_hostname, strlen(server_hostname),
+                                   "idna", "strict");
```
Comment (vstinner) [inline Modules/_ssl.c:583]: Hum, this change doesn't seem to be directely related to MemoryBIO, or did I miss something?

File: Modules/_ssl.c [hunk 5]
```diff
@@ -2015,6 +2127,12 @@ If the TLS handshake is not yet complete, None is returned");
 static PyGetSetDef ssl_getsetlist[] = {
     {"context", (getter) PySSL_get_context,
                 (setter) PySSL_set_context, PySSL_set_context_doc},
+    {"server_side", (getter) PySSL_get_server_side, NULL,
+                    PySSL_get_server_side_doc},
+    {"server_hostname", (getter) PySSL_get_server_hostname, NULL,
+                        PySSL_get_server_hostname_doc},
```
Comment (vstinner) [inline Modules/_ssl.c:2157]: Are these changes related to MemoryBIO? Maybe they should be backported in a different commit or PR?

File: Doc/library/ssl.rst [hunk 6]
```diff
@@ -1688,6 +1774,143 @@ to be aware of:
         except ssl.SSLWantWriteError:
             select.select([], [sock], [])
 
+.. seealso::
+
+   The :mod:`asyncio` module supports :ref:`non-blocking SSL sockets
```
Comment (alex) [inline Doc/library/ssl.rst]: The `:mod:` link probably doesn't work on py2 :-)

File: Doc/library/ssl.rst (hunk 1 above)
Comment (alex) [reply Doc/library/ssl.rst]: This should be fixed before it's merged :-)

File: Doc/library/ssl.rst (hunk 2 above)
Comment (alex) [reply Doc/library/ssl.rst:1820]: 👍

File: Lib/ssl.py [hunk 7]
```diff
@@ -715,17 +848,7 @@ def send(self, data, flags=0):
...
-            except SSLError as x:
-                if x.args[0] == SSL_ERROR_WANT_READ:
-                    return 0
-                elif x.args[0] == SSL_ERROR_WANT_WRITE:
-                    return 0
-                else:
-                    raise
-            else:
-                return v
+            return self._sslobj.write(data)
```
Comment (alex) [inline Lib/ssl.py]: Were these changes in my original patch?

File: Modules/_ssl.c (hunk 4 above)
Comment (alex) [reply Modules/_ssl.c:583]: I think it's related to teh weakrefing, but my memory is foggy :-(

**Review by alex — COMMENTED** [review]
It's been a while since I looked at this patch, but from an initial glance it looks like what I expected.

**Comment by alex** [comment]
Tests appear to be hanging; these suck to debug. Usually there's a hidden exception coming from one of the non-main threads.

File: Doc/library/ssl.rst (hunk 6 above)
Comment (vstinner) [reply Doc/library/ssl.rst]: Oh. I removed this note :-) Inaccurate in Python2 doc.

File: Doc/library/ssl.rst (hunk 1 above)
Comment (vstinner) [reply Doc/library/ssl.rst]: fixed

File: Doc/library/ssl.rst (hunk 2 above)
Comment (vstinner) [reply Doc/library/ssl.rst:1820]: done

File: Lib/ssl.py (hunk 3 above)
Comment (vstinner) [reply Lib/ssl.py]: fixed

File: Lib/ssl.py (hunk 7 above)
Comment (vstinner) [reply Lib/ssl.py]: If you don't know, I will just revert it.

File: Modules/_ssl.c (hunk 5 above)
Comment (vstinner) [reply Modules/_ssl.c:2157]: @alex replied above: it's related a weakref change.

**Comment by vstinner** [comment]
I rebased my change and addressed @alex and my comments. I removed [WIP] from the PR title, it's now ready for a real review :-)

**Comment by alex** [comment]
Tests still appear to be hanging, so that's going to be a blocker, I'll review for everything else.

File: Modules/_ssl.c [hunk 8]
```diff
@@ -566,8 +592,17 @@ newPySSLSocket(PySSLContext *sslctx, PySocketSockObject *sock,
...
-    SSL_set_app_data(self->ssl,self);
-    SSL_set_fd(self->ssl, Py_SAFE_DOWNCAST(sock->sock_fd, SOCKET_T, int));
+    SSL_set_app_data(self->ssl, self);
+    if (sock) {
+        SSL_set_fd(self->ssl, Py_SAFE_DOWNCAST(sock->sock_fd, SOCKET_T, int));
+    } else {
+        /* BIOs are reference counted and SSL_set_bio borrows our reference.
+         * To prevent a double free in memory_bio_dealloc() we need to take an
+         * extra reference here. */
+        CRYPTO_add(&inbio->bio->references, 1, CRYPTO_LOCK_BIO);
```
Comment (alex) [inline Modules/_ssl.c]: These need to be converted to calls to `BIO_up_ref`, for OpenSSL 1.1.0 compat

**Review by alex — COMMENTED** [review]
Patch looks fine (and largely the same as my original one!). If you'd like help tracking down the hang in test_ssl, let me know.

**Comment by alex** [comment]
Sigh, it took quite a while, but the answer is that `svn.python.org` doesn't behave the exact way these tests expect. There's a reason we ported all the rest of the tests away from it :-(

Switching to `REMOTE_HOST` fixes `test_read_write_data`, but it looks like `test_handshake` is really only happy with the in-memoryserver that the py3k versions of these tests use.

Here's the diff I used to do all this testing:

```diff
diff --git a/Lib/test/test_ssl.py b/Lib/test/test_ssl.py
index 4e34734662..c99db781c1 100644
--- a/Lib/test/test_ssl.py
+++ b/Lib/test/test_ssl.py
@@ -1757,17 +1757,17 @@ class NetworkedBIOTests(unittest.TestCase):
         return ret
 
     def test_handshake(self):
-        with support.transient_internet("svn.python.org"):
+        with support.transient_internet(REMOTE_HOST):
             sock = socket.socket(socket.AF_INET)
-            sock.connect(("svn.python.org", 443))
+            sock.connect((REMOTE_HOST, 443))
             incoming = ssl.MemoryBIO()
             outgoing = ssl.MemoryBIO()
             ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
             ctx.verify_mode = ssl.CERT_REQUIRED
-            ctx.load_verify_locations(SVN_PYTHON_ORG_ROOT_CERT)
+            ctx.load_verify_locations(REMOTE_ROOT_CERT)
             if ssl.HAS_SNI:
                 ctx.check_hostname = True
-                sslobj = ctx.wrap_bio(incoming, outgoing, False, 'svn.python.org')
+                sslobj = ctx.wrap_bio(incoming, outgoing, False, REMOTE_HOST)
             else:
                 ctx.check_hostname = False
                 sslobj = ctx.wrap_bio(incoming, outgoing, False)
@@ -1786,9 +1786,9 @@ class NetworkedBIOTests(unittest.TestCase):
             sock.close()
 
     def test_read_write_data(self):
-        with support.transient_internet("svn.python.org"):
+        with support.transient_internet(REMOTE_HOST):
             sock = socket.socket(socket.AF_INET)
-            sock.connect(("svn.python.org", 443))
+            sock.connect((REMOTE_HOST, 443))
             incoming = ssl.MemoryBIO()
             outgoing = ssl.MemoryBIO()
             ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
```

**Comment by vstinner** [comment]
@alex: I'm unable to reproduce the failure on my Fedora 25, so I tried your patch to see if it's going better on the CI.

File: Modules/_ssl.c (hunk 8 above)
Comment (vstinner) [reply Modules/_ssl.c]: @alex, @tiran: Honestly, I don't know how to do that. I only tried to convert an "old" patch to a more recent PR, but feel free to take the ownership of the PR and make your own changes if you want ;-) (We just have to sync, maybe on IRC?)

**Comment by alex** [comment]
FWIW, I had no problem reproducing on macOS. What version of OpenSSL does Fedora 25 use?

**Comment by vstinner** [comment]
Fedora 25 uses OpenSSL 1.0.2k.

**Comment by tiran** [comment]
You have to use ``BIO_up_ref()`` instead of ``CRYPTO_add``. 3.6 has:

```
diff --git a/Modules/_ssl.c b/Modules/_ssl.c
index 8202fd0262..defb50ed4a 100644
--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -171,6 +171,12 @@ static X509 *X509_OBJECT_get0_X509(X509_OBJECT *x)
     return x->data.x509;
 }
 
+static int BIO_up_ref(BIO *b)
+{
+    CRYPTO_add(&b->references, 1, CRYPTO_LOCK_BIO);
+    return 1;
+}
+
 static STACK_OF(X509_OBJECT) *X509_STORE_get0_objects(X509_STORE *store) {
     return store->objs;
 }
@@ -599,8 +605,8 @@ newPySSLSocket(PySSLContext *sslctx, PySocketSockObject *sock,
         /* BIOs are reference counted and SSL_set_bio borrows our reference.
          * To prevent a double free in memory_bio_dealloc() we need to take an
          * extra reference here. */
-        CRYPTO_add(&inbio->bio->references, 1, CRYPTO_LOCK_BIO);
-        CRYPTO_add(&outbio->bio->references, 1, CRYPTO_LOCK_BIO);
+        BIO_up_ref(inbio->bio);
+        BIO_up_ref(outbio->bio);
         SSL_set_bio(self->ssl, inbio->bio, outbio->bio);
     }
     mode = SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER;
```

**Comment by vstinner** [comment]
Sadly, MemoryBIO will have to wait for Python 2.7.15 which has no schedule yet :-(

**Comment by tiran** [comment]
The backport has a bug. ``sslobj.unwrap`` call in ``NetworkedBIOTests.test_handshake`` blocks indefinitely. The test is only executed with ``-u network`` resource enabled..

**Comment by vstinner** [comment]
```
./python -m test -v -u all -m test_handshake test_ssl
```

This command loops on the "SSLSyscallError: Some I/O error occurred (_ssl.c:2097)" error while running SSLObject.unwrap(): "self.ssl_io_loop(sock, incoming, outgoing, sslobj.unwrap)".
